src/ShallowWaterIntegrators.jl

```diff
@@ -0,0 +1,141 @@
+function assemble_rhs_vector(A, B, dc)
+  # short cut routine to assemble a domain contribution into a vector
+  # A:  test function
+  # B:  trial function
+  # dc: domain contribution
+  data  = Gridap.FESpaces.collect_cell_vector(B, dc)
+  assem = SparseMatrixAssembler(A, B)
+  rhs   = assemble_vector(assem, data)
+end
+
+function shallow_water_explicit(model, order, Ω, dΩ, dω, qₖ, wₖ, f, g, h₁, u₁, hₘ, uₘ, RTMM, L2MM, dt, leap_frog, τ, P, Q, U, V, R, S)
+  # energetically balanced explicit second order shallow water solver
+  # reference: eqns (21-24) of
+  # https://github.com/BOM-Monash-Collaborations/articles/blob/main/energetically_balanced_time_integration/EnergeticallyBalancedTimeIntegration_SW.tex
+  #
+  # f          : coriolis force (field)
+  # g          : gravity (constant)
+  # h₁         : fluid depth at current time level
+  # u₁         : fluid velocity at current time level
+  # hₘ         : fluid depth at previous time level (for leap-frogging the first step)
+  # uₘ         : fluid velocity at previous time level (for leap-frogging the first step)
+  # RTMM       : H(div) mass matrix, ∫β⋅βdΩ, ∀β∈ H(div,Ω)
+  # L2MM       : L² mass matrix, ∫γγdΩ, ∀γ∈ L²(Ω)
+  # dt         : time step
+  # leap_frog  : do leap frog time integration for the first step (boolean)
+  # order      : polynomial order
+  # dΩ         : measure of the elements
+
+  u = get_trial_fe_basis(U)
+  v = get_fe_basis(V)
+  p = get_trial_fe_basis(P)
+  q = get_fe_basis(Q)
+  n = get_normal_vector(model)
+
+  # explicit step for provisional velocity, uₚ
+  dt1 = dt
+  if leap_frog
+    dt1 = 2.0*dt
+  end
+
+  # 1.1: the mass flux
+  b₁(v)  = ∫(v⋅u₁*h₁)*dΩ
+  rhs1   = assemble_rhs_vector(U, V, b₁(v))
+  op     = AffineFEOperator(U, V, RTMM, rhs1)
+  F      = solve(op)
```

Comment:
Ok. This is a critical line from the performance POV. Each time we call `solve` an LU factorizatoin of `RTMM` is computed under the hood. `RTMM` and `L2MM` are not time-dependent, thus this is not needed. 

Do not use `solve`. Instead, compute once `lu` factorization

https://github.com/gridapapps/GridapGeosciences.jl/blob/master/test/WaveEquationCubedSphereTests.jl#L80

and then do forward & backward substitution at each step

https://github.com/gridapapps/GridapGeosciences.jl/blob/master/test/WaveEquationCubedSphereTests.jl#L31

Also note that in that line the `res` vector is re-used over and over again. Your `solve` call also allocates a new vector on each call.